File: packages/asdl-slots/src/asdl_slots/cli/slot/gt/free_stack.py
from __future__ import annotations

import click

from asdl_core import get_console
from asdl_core.clinkr.exit import ClinkrExit
from asdl_core.clinkr.models import ClinkrModel
from asdl_core.clinkr.operation import clinkr_operation
from asdl_core.git.types import DetachedHead
from asdl_core.git.types import GitCommandFailure as GitFailure
from asdl_core.gt.types import GtCommandFailure
from asdl_slots.cli.slot.free import FreedSlot
from asdl_slots.cli.slot.gt.context import load_slot_gt_context
from asdl_slots.cli.slot.gt.stack_walk import collect_stack_branches
from asdl_slots.inventory import SlotMatch, build_slot_inventory
from asdl_slots.lifecycle import SlotFreeOutcome, SlotLifecycleFailure, free_slots
from asdl_slots.repo_context import NoRepoSentinel


class SlotGtFreeStackRequest(ClinkrModel):
    pass


class SlotGtFreeStackResult(ClinkrModel):
    current_branch: str
    trunk_branch: str
    freed: tuple[FreedSlot, ...]
    noop_reason: str | None


def render_slot_gt_free_stack(result: SlotGtFreeStackResult) -> None:
    console = get_console()
    if result.noop_reason == "on_trunk":
        console.print(
            f"[dim]On trunk ([yellow]{result.trunk_branch}[/yellow]); nothing to free.[/dim]"
        )
        return
    if result.noop_reason == "no_slots":
        console.print("[dim]No slots in stack to free.[/dim]")
        return
    for entry in result.freed:
        console.print(
            f"[green]✓[/green] Freed [bold cyan]{entry.slot_name}[/bold cyan] "
            f"([yellow]{entry.branch_name}[/yellow])"
        )
        console.print(
            f"  Worktree kept at [dim]{entry.worktree_path}[/dim]; detached HEAD at trunk"
        )


@clinkr_operation(
    name="free-stack",
    help=(
        "Release every slot in the current Graphite stack except the slot at the current branch."
    ),
    human_renderer=render_slot_gt_free_stack,
)
def run_gt_free_stack(
    ctx: click.Context, request: SlotGtFreeStackRequest
) -> ClinkrExit[SlotGtFreeStackResult]:
    gt_ctx = load_slot_gt_context(ctx)
    if isinstance(gt_ctx, NoRepoSentinel):
        raise ClinkrExit.failure(error_type="not_in_repo", message=gt_ctx.message)

    slots_ctx = gt_ctx.slots
    current_result = slots_ctx.git.get_current_branch(slots_ctx.repo.root)
    if isinstance(current_result, GitFailure):
        raise ClinkrExit.failure(
            error_type="git_current_branch_failed",
            message=current_result.message,
        )
    if isinstance(current_result, DetachedHead):
        raise ClinkrExit.failure(
            error_type="detached_head",
            message=f"HEAD at {slots_ctx.repo.root} is detached. Check out a branch first.",
        )
    current = current_result

    trunk_result = gt_ctx.gt.trunk(slots_ctx.repo.root)
    if isinstance(trunk_result, GtCommandFailure):
        raise ClinkrExit.failure(error_type="gt_trunk_failed", message=trunk_result.message)
    trunk = trunk_result

    if current == trunk:
        return ClinkrExit.ok(
            SlotGtFreeStackResult(
                current_branch=current,
                trunk_branch=trunk,
                freed=(),
                noop_reason="on_trunk",
            )
        )

    inventory = build_slot_inventory(
        slots_ctx.git,
        main_repo_root=slots_ctx.repo.main_repo_root,
    )
    if inventory.pool_size == 0:
        raise ClinkrExit.failure(
            error_type="pool_empty",
            message="No managed slots configured. Run `slot init --size N` first.",
        )

    stack_result = gt_ctx.gt.stack(slots_ctx.repo.root)
    if isinstance(stack_result, GtCommandFailure):
        raise ClinkrExit.failure(error_type="gt_stack_failed", message=stack_result.message)
    stack = stack_result

    stack_branches = collect_stack_branches(stack, current=current, trunk=trunk)

    seen: set[str] = set()
    targets: list[str] = []
    for branch in stack_branches:
        # `collect_stack_branches` already excludes current and trunk;
        # double-check here in case a future refactor relaxes that.
        if branch == current or branch == trunk:
            continue
        match = inventory.find_by_branch(branch)
        if not isinstance(match, SlotMatch):
            continue
        slot_name = match.record.slot_name
        if slot_name in seen:
            continue
        seen.add(slot_name)
        targets.append(slot_name)

    if not targets:
        return ClinkrExit.ok(
            SlotGtFreeStackResult(
                current_branch=current,
                trunk_branch=trunk,
                freed=(),
                noop_reason="no_slots",
            )
        )

    outcome = free_slots(slots_ctx, tuple(targets), trunk_branch=trunk)
    if isinstance(outcome, SlotLifecycleFailure):
        raise ClinkrExit.failure(error_type=outcome.error_type, message=outcome.message)
    assert isinstance(outcome, SlotFreeOutcome)

    return ClinkrExit.ok(
        SlotGtFreeStackResult(
            current_branch=current,
            trunk_branch=trunk,
            freed=tuple(
                FreedSlot(
                    slot_name=entry.slot_name,
                    branch_name=entry.branch_name,
                    worktree_path=str(entry.worktree_path),
                )
                for entry in outcome.freed
            ),
            noop_reason=None,
        )
    )

File: ts/packages/slot/src/operations/gt/free-stack.ts
import { failure, ok } from "@asdl/clinkr";
import { z } from "zod";

import type { RepoSlotContext, SlotCliContext } from "../../context.ts";
import { buildSlotInventory, findByBranch, poolSize } from "../../inventory.ts";
import { executeFreePlan, planFreeSlots } from "../../lifecycle/free.ts";
import { collectStackBranches } from "./stack-walk.ts";

const freedSlotSchema = z.object({ slot_name: z.string(), branch_name: z.string(), worktree_path: z.string() });

export const gtFreeStackRequestSchema = z.object({
	downstack: z.boolean().default(false).describe("Free only ancestor/downstack slots."),
});

export const gtFreeStackResultSchema = z.object({
	current_branch: z.string(),
	trunk_branch: z.string(),
	freed: z.array(freedSlotSchema),
	noop_reason: z.union([z.literal("on_trunk"), z.literal("no_slots")]).nullable(),
	downstack: z.boolean(),
});

export type GtFreeStackRequest = z.infer<typeof gtFreeStackRequestSchema>;
export type GtFreeStackResult = z.infer<typeof gtFreeStackResultSchema>;

export async function runGtFreeStack(ctx: SlotCliContext, request: GtFreeStackRequest) {
	if (ctx.repo.type !== "repo") return failure(ctx.repo.errorType, ctx.repo.message);
	if (ctx.gt === undefined) return failure("gt_gateway_missing", "Graphite gateway is not available for slot gt commands.");
	const repoCtx: RepoSlotContext = { ...ctx, repo: ctx.repo };
	const currentResult = await repoCtx.git.getCurrentBranch(repoCtx.repo.root);
	if (currentResult.type === "failure") return failure("git_current_branch_failed", currentResult.failure.message);
	if (currentResult.type === "detached") return failure("detached_head", `HEAD at ${repoCtx.repo.root} is detached. Check out a branch first.`);
	const trunkResult = await ctx.gt.trunk(repoCtx.repo.root);
	if (trunkResult.type === "failure") return failure("gt_trunk_failed", trunkResult.failure.message);
	if (currentResult.branch === trunkResult.branch) return ok(buildResult({ current: currentResult.branch, trunk: trunkResult.branch, freed: [], noopReason: "on_trunk", downstack: request.downstack }));
	const inventory = await buildSlotInventory(repoCtx.git, { mainRepoRoot: repoCtx.repo.mainRepoRoot });
	if (poolSize(inventory) === 0) return failure("pool_empty", "No managed slots configured. Run `slot init --size N` first.");
	const stackResult = await ctx.gt.stack(repoCtx.repo.root);
	if (stackResult.type === "failure") return failure("gt_stack_failed", stackResult.failure.message);
	if (stackResult.type === "untracked_branch") return failure("gt_untracked_branch", "current branch is not tracked by Graphite — run `gt track` first");
	const branches = collectStackBranches(stackResult.stack, { current: currentResult.branch, trunk: trunkResult.branch, downstackOnly: request.downstack, includeCurrent: false });
	const targets: string[] = [];
	const seen = new Set<string>();
	for (const branch of branches) {
		if (branch === currentResult.branch || branch === trunkResult.branch) continue;
		const match = findByBranch(inventory, branch);
		if (match?.kind !== "slot") continue;
		if (seen.has(match.record.slotName)) continue;
		seen.add(match.record.slotName);
		targets.push(match.record.slotName);
	}
	if (targets.length === 0) return ok(buildResult({ current: currentResult.branch, trunk: trunkResult.branch, freed: [], noopReason: "no_slots", downstack: request.downstack }));
	const plan = await planFreeSlots(repoCtx, targets, { trunkBranch: trunkResult.branch });
	if (plan.type === "failure") return failure(plan.failure.error_type, plan.failure.message);
	const executed = await executeFreePlan(repoCtx, plan.outcome);
	if (executed.type === "failure") return failure(executed.failure.error_type, executed.failure.message);
	return ok(buildResult({ current: currentResult.branch, trunk: trunkResult.branch, freed: executed.outcome.freed, noopReason: null, downstack: request.downstack }));
}

export function renderGtFreeStack(result: GtFreeStackResult): string {
	if (result.noop_reason === "on_trunk") return `On trunk (${result.trunk_branch}); nothing to free.`;
	if (result.noop_reason === "no_slots") return `No slots in stack to free${result.downstack ? " (downstack only)" : ""}.`;
	return result.freed.map((entry) => `Freed ${entry.slot_name} (${entry.branch_name})\n  Worktree kept at ${entry.worktree_path}; detached HEAD at trunk`).join("\n");
}

function buildResult(options: { current: string; trunk: string; freed: readonly { slot_name: string; branch_name: string; worktree_path: string }[]; noopReason: "on_trunk" | "no_slots" | null; downstack: boolean }): GtFreeStackResult {
	return { current_branch: options.current, trunk_branch: options.trunk, freed: options.freed.map((entry) => ({ slot_name: entry.slot_name, branch_name: entry.branch_name, worktree_path: entry.worktree_path })), noop_reason: options.noopReason, downstack: options.downstack };
}
